### ISSUE TITLE
[kfctl] Create kfctl CLI 

### ISSUE THREAD

**Issue by kunmingg** [kubeflow/kubeflow#1419]
Should cover same user experience as https://github.com/kubeflow/kubeflow/blob/master/scripts/kfctl.sh

Can use https://github.com/spf13/cobra

Available APIs here:
https://github.com/kubeflow/kubeflow/blob/master/bootstrap/cmd/bootstrap/app/ksServer.go

**Comment by TheJaySmith** [comment]
Hello,

Go works too. Either way, do we have someone taking the action item?

**Comment by kunmingg** [comment]
No, would be great if you have time for it?

**Comment by TheJaySmith** [comment]
Sure. I am working on a CLI for some of the WeaveWorks Flux integration and it doesn't make sense to have two CLI's within the same product. My GoLang is rusty but what better way to improve it.

**Comment by kkasravi** [comment]
@TheJaySmith take a look at [arena](https://github.com/kubeflow/arena/tree/master/cmd/arena/commands) which uses spf13/cobra and is a kubeflow CLI with a number of commands. You could possibly add the kfctl sub-commands there. I think at some point these CLI's would be combined (kfctl and arena).
@kunmingg @jlewi feel free to weigh in.

I'm doing some related work and may ping you on this stuff.

**Comment by swiftdiaries** [comment]
I'd like to take this up if someone isn't already working on it. Also, would appreciate some direction as to where it would go? Part of arena or in kubeflow/components?

**Comment by jlewi** [comment]
The go binary is here:
https://github.com/kubeflow/kubeflow/tree/master/bootstrap/

This is where we should add additional logic to make the binary a replacement for kfctl.sh.

@swiftdiaries I think a good place to start would be to just go through the commands that we execute with kfctl.sh and add entry points any missing logic; e.g.

```
kfctl init ...
kfctal generate platform ...
kfctl generate k8s ...
kfctl apply platform ...
kfctl apply k8s ...
```

**Comment by jlewi** [comment]
@TheJaySmith if your working on this can you please clarify what exactly you are working on and when you plan on delivering it?

**Comment by chrisheecho** [comment]
/assign @TheJaySmith 
Jay it seems like you are working on it? just want to make it official

**Comment by TheJaySmith** [comment]
@chrisheecho yes. I have been a tad delayed but plan on having a deliverable by mid next week. Keep me to it

**Comment by jlewi** [comment]
@TheJaySmith are you still working on this? I think @ashahba and @kkasravi expressed interest in working on this?

Do we think we can get this done in time for 0.4 (which we will cut on 12/17) if not should we punt this to 0.5?

**Comment by TheJaySmith** [comment]
Yep, sorry, just got a little delayed. I will upload a README. Let's get this for 0.4!

**Comment by jlewi** [comment]
Moving this to 0.5.

@kkasravi Looks like #2016 has been pending for a while.

**Comment by kkasravi** [comment]
@jlewi will open a new PR early next week that has our [agreed upon design](https://github.com/kubeflow/kubeflow/tree/master/kfctl#proposed-design)

**Comment by jlewi** [comment]
Thanks @kkasravi 

/assign @kkasravi 
/unassign thejaysmith

**Comment by kkasravi** [comment]
closed by #2294
